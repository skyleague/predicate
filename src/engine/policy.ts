import { collect, stack } from '@skyleague/axioms'
import type { IsEmptyObject, Simplify } from '@skyleague/axioms/types'
import packageJSON from '../../package.json' with { type: 'json' }
import type { FactsFomExprs } from './operator.js'
import type { Expression, ExpressionReturnType } from './types.js'

const { version } = packageJSON

export class EvaluationContext {
    public evaluated = new WeakMap<object, unknown>()
    public state: Partial<Record<PropertyKey, unknown>> = {}
    public input: Record<string, unknown> = {}
    public scope: unknown = {}
    public stack: (readonly [EvaluationContext['evaluated'], EvaluationContext['state'], unknown])[]

    public constructor(input: Record<string, unknown>) {
        this.input = input
        this.stack = [[this.evaluated, this.state, this.scope]]
    }

    public evaluate<O>(expr: Expression<O>): O {
        if (this.evaluated.has(expr)) {
            return this.evaluated.get(expr) as O
        }

        const input = expr.dependsOn.map((d) => this.evaluate(d))

        const result = expr.fn(input, this)
        this.evaluated.set(expr, result)
        return result
    }

    public withScope<T>(scope: unknown, fn: () => T) {
        this.push(scope)
        const result = fn()
        this.pop()
        return result
    }

    protected push(scope: unknown) {
        const top = [new WeakMap<object, unknown>(), {}, scope] as const
        this.stack.push(top)
        this.evaluated = top[0]
        this.state = top[1]
        this.scope = top[2]
    }

    protected pop() {
        this.stack.pop()
        const top = this.stack.at(-1) ?? [this.evaluated, this.state, this.scope]
        this.evaluated = top[0]
        this.state = top[1]
        this.scope = top[2]
    }
}

export interface _EvaluationContext {
    evaluated: WeakMap<object, unknown>
    state: Partial<Record<PropertyKey, unknown>>
    stack: [_EvaluationContext['evaluated'], _EvaluationContext['state'], unknown][]
    scope: unknown
    evaluate<T>(expr: Expression<T>): T
    withScope: <T>(y: unknown, fn: () => T) => T
}

function* collapseExpression(root: Expression[], seen = new WeakSet()) {
    const nodes = stack(root)
    for (const n of nodes) {
        if (!seen.has(n)) {
            yield n
            nodes.push(n.dependsOn)
            seen.add(n)
        }
    }
}
type InferFactName<Expr, k> = Expr extends { name: string } ? Expr['name'] : k

type FilterFactExpressions<Facts, K extends keyof Facts = keyof Facts> = Facts[K] extends {
    _type: 'fact'
}
    ? InferFactName<Facts[K], K>
    : never

type _InputFromName<Names extends PropertyKey, Facts> = {
    [K in Names]: ExpressionReturnType<Extract<Facts[keyof Facts], { name: K }>>
}
type _InputFromExpressions<Facts> = _InputFromName<
    FilterFactExpressions<Facts> extends PropertyKey ? FilterFactExpressions<Facts> : never,
    Facts
>

type _FactsFomExprs<Facts> = Facts extends unknown[] ? Facts[number] : Facts
export type InputFromExpressions<Facts extends Record<string, unknown>> = Simplify<
    _InputFromExpressions<{ [K in keyof Facts as FilterFactExpressions<Facts, K>]: Facts[K] }> &
        _InputFromExpressions<{ [K in keyof Facts]: _FactsFomExprs<FactsFomExprs<Facts[K]>> }>
>

export type OutputFromFacts<Facts extends Record<string, unknown>> = Simplify<{
    [k in keyof Facts]: ExpressionReturnType<Facts[k]>
}>

export interface Policy<I, O> {
    evaluate: IsEmptyObject<I> extends true ? () => { input: I; output: O } : (x: I) => { input: I; output: O }
    expr: () => unknown
}

export function $policy<Facts extends Record<string, Expression>>(
    expressions: Facts,
): Policy<InputFromExpressions<Facts>, OutputFromFacts<Facts>> {
    const facts = Object.entries(expressions).map(([name, e]) => {
        if (e.name === undefined) {
            return Object.assign(e, { name })
        }
        return e
    })

    const allNodes = collect(collapseExpression(facts))
    const inputNodes = allNodes.filter(
        (f: (typeof allNodes)[number]): f is (typeof allNodes)[number] & { name: string } =>
            f._type === 'fact' && f.name !== undefined,
    )
    const outputNodes = allNodes.filter((f) => f._type !== 'fact' && f.name !== undefined)

    const properties = Object.fromEntries(inputNodes.map((e) => [e.name, e.expr('definition')]))
    const outputExpression = Object.fromEntries(outputNodes.map((f) => [f.name, f.expr('expression')]))
    return {
        evaluate: ((input: Record<string, unknown>) => {
            const ctx = new EvaluationContext(input)

            for (const fact of facts) {
                const result = ctx.evaluate(fact)
                if (fact.name !== undefined) {
                    ctx.state[fact.name] = result
                }
            }

            return { input: input, output: ctx.state }
        }) as Policy<InputFromExpressions<Facts>, OutputFromFacts<Facts>>['evaluate'],
        expr: () => {
            const input = properties
            return {
                meta: {
                    version,
                },
                input,
                output: outputExpression,
            }
        },
    }
}
