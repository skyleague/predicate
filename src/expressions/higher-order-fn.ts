import type { EvaluationContext } from '../engine/policy.js'
import {
    type AsExpression,
    type Expression,
    type ExpressionTypeOfLiteral,
    type InferExpressionType,
    type InputExpression,
    type LiteralOr,
    type ValueExpression,
    fromLiteral,
} from '../engine/types.js'
import type { ValueItemExpr } from '../json/jsonexpr.type.js'

import { JSONPath, type JSONPathValue } from '@skyleague/jsonpath'

import { inspect } from 'node:util'

// biome-ignore lint/suspicious/noExplicitAny: this is needed for greedy matching
export interface Value<O> extends InputExpression<O, any, [], ValueItemExpr> {
    _type: 'value'
}
function value<T, P extends string | undefined = undefined>(path?: P): P extends string ? Value<JSONPathValue<T, P>> : Value<T> {
    if (path === undefined) {
        return {
            _type: 'value',
            dependsOn: [],
            fn: (_: unknown, ctx: EvaluationContext) => ctx.scope,
            expr: () => ({ value: '$' }),
            [inspect.custom]() {
                return 'x'
            },
        } as unknown as P extends string ? Value<JSONPathValue<T, P>> : Value<T>
    }
    return {
        _type: 'value',
        dependsOn: [],
        fn: (_: unknown, ctx: EvaluationContext) => JSONPath.get(ctx.scope, path),
        expr: () => ({ value: path.toString() }),
        [inspect.custom]() {
            return `x(${path})`
        },
    } as unknown as P extends string ? Value<JSONPathValue<T, P>> : Value<T>
}

// biome-ignore lint/suspicious/noExplicitAny: this is needed for greedy matching
type ValueFn<I extends any[]> = <P extends string | undefined = undefined>(
    path?: P,
) => P extends string ? Value<JSONPathValue<I[number], P>> : Value<I[number]>

// biome-ignore lint/suspicious/noExplicitAny: this is needed for greedy matching
export type ValueItem<I extends any[]> = ValueFn<I> & Value<I[number]>

// biome-ignore lint/suspicious/noExplicitAny: this is needed for greedy matching
export function $value<T extends any[]>(): ValueItem<T> {
    const val = value()
    // biome-ignore lint/suspicious/noExplicitAny: this is needed for greedy matching
    return Object.assign((x: any) => value(x), val) as ValueItem<T>
}

export const $map = Object.assign(
    // biome-ignore lint/suspicious/noExplicitAny: this is needed for greedy matching
    <O extends LiteralOr<any>, Expr extends LiteralOr<any[]>>(
        xs: Expr,
        transform: (value: ValueItem<ExpressionTypeOfLiteral<Expr>>) => AsExpression<O>,
    ): ValueExpression<ExpressionTypeOfLiteral<O>[], [AsExpression<Expr>], InferExpressionType<ExpressionTypeOfLiteral<O>>> => {
        const _xs = fromLiteral(xs)
        const _value = $value<ExpressionTypeOfLiteral<Expr>>()
        const _transform = transform(_value)
        return {
            fn: ([ys], ctx) => {
                return ys.map((x) => ctx.withScope(x, () => ctx.evaluate(_transform)))
            },
            dependsOn: [_xs],
            expr: (mode) => ({
                map: [_xs.expr(mode), _transform.expr(mode)],
            }),
            [inspect.custom]() {
                return `$map(${_xs[inspect.custom]?.() ?? ''}, (x) => ${_transform[inspect.custom]?.() ?? ''})`
            },
        } as ValueExpression<ExpressionTypeOfLiteral<O>[], [AsExpression<Expr>], InferExpressionType<ExpressionTypeOfLiteral<O>>>
    },
    // biome-ignore lint/suspicious/noExplicitAny: this is needed for greedy matching
    { operator: 'map', symbol: '$map', parse: (xs: any, transform: any) => $map(xs, () => transform) } as const,
)

export const $filter = Object.assign(
    // biome-ignore lint/suspicious/noExplicitAny: this is needed for greedy matching
    <Expr extends LiteralOr<any[]>>(
        xs: Expr,
        predicate: (value: ValueItem<ExpressionTypeOfLiteral<Expr>>) => Expression<boolean>,
    ): ValueExpression<
        ExpressionTypeOfLiteral<Expr>,
        [AsExpression<Expr>],
        InferExpressionType<ExpressionTypeOfLiteral<Expr>>
    > => {
        const _xs = fromLiteral(xs)
        const _value = $value<ExpressionTypeOfLiteral<Expr>>()
        const _transform = predicate(_value)
        return {
            fn: ([ys], ctx) => {
                return ys.filter((x) => ctx.withScope(x, () => ctx.evaluate(_transform)))
            },
            dependsOn: [_xs],
            expr: (mode) => ({
                filter: [_xs.expr(mode), _transform.expr(mode)],
            }),
            [inspect.custom]() {
                return `$filter(${_xs[inspect.custom]?.() ?? ''}, (x) => ${_transform[inspect.custom]?.() ?? ''})`
            },
        } as ValueExpression<
            ExpressionTypeOfLiteral<Expr>,
            [AsExpression<Expr>],
            InferExpressionType<ExpressionTypeOfLiteral<Expr>>
        >
    },
    // biome-ignore lint/suspicious/noExplicitAny: this is needed for greedy matching
    { operator: 'filter', symbol: '$filter', parse: (xs: any, predicate: any) => $filter(xs, () => predicate) } as const,
)
