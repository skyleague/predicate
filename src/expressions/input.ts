import type {
    FactExpression,
    DefinitionType,
    Expression,
    InputExpression,
    InferExpressionType,
    LiteralExpression,
} from '../engine/types.js'
import type { FromExpr } from '../json/jsonexpr.type.js'

import { type JSONPathValue, JSONPath } from '@skyleague/jsonpath'
import type { Schema } from '@skyleague/therefore'

import { inspect } from 'node:util'

export interface Fact<T, Name extends string> extends FactExpression<T> {
    name: Name
    dependsOn: []
    _type: 'fact'
}

export function $fact<T, Name extends string>(schema: Pick<Schema<T>, 'schema' | 'is'>, name: Name): Fact<T, Name> {
    return {
        _type: 'fact',
        dependsOn: [],
        name: name,
        fn: ((_: unknown, ctx: { input: Record<Name, T> }) => {
            return ctx.input[name]
        }) as Fact<T, Name>['fn'],
        expr: (definition: DefinitionType) =>
            (definition === 'definition' ? schema.schema : { fact: name.toString() }) as InferExpressionType<T>,
        [inspect.custom]() {
            return `$fact({schema: ${JSON.stringify(schema.schema)}}, "${name}")`
        },
    }
}
export interface From<I, O, DependsOn extends Expression[]> extends InputExpression<O, I, DependsOn, FromExpr> {
    _type: 'value'
}

export function $from<T, Name extends string>(fact: Fact<T, Name>): From<T, JSONPathValue<T, '$'>, [Fact<T, Name>]>
export function $from<T, P extends string, Name extends string>(
    fact: Fact<T, Name>,
    path: P
): From<T, JSONPathValue<T, P>, [Fact<T, Name>]>
export function $from<T, P extends string, Name extends string>(
    fact: Fact<T, Name>,
    path: P = '$' as P
): From<T, JSONPathValue<T, P>, [Fact<T, Name>]> {
    return {
        _type: 'value',
        dependsOn: [fact],
        fn: ((_: unknown, ctx: { input: Record<Name, T> }) => JSONPath.get(ctx.input[fact.name], path)) as From<
            T,
            JSONPathValue<T, P>,
            [Fact<T, Name>]
        >['fn'],
        expr: (_definition) => ({ from: [fact.name, path.toString()] as const }),
        [inspect.custom]() {
            return `$from(${fact[inspect.custom]?.() ?? `"${fact.name}"`}${path === '$' ? '' : `, "${path}"`})`
        },
    }
}

export function $literal<O>(x: O): LiteralExpression<O> {
    return {
        fn: () => x,
        dependsOn: [],
        expr: () => x as InferExpressionType<O>,
        _type: 'literal',
        [inspect.custom]() {
            return `$literal(${inspect(x)})`
        },
    }
}
