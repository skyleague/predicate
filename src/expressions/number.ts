import { operator } from '../engine/operator.js'

export const $add = operator({ operator: '+', symbol: '$add', fn: ([a, b]: [number, number]) => a + b })
export const $subtract = operator({ operator: '-', symbol: '$subtract', fn: ([a, b]: [number, number]) => a - b })
export const $divide = operator({ operator: '/', symbol: '$divide', fn: ([a, b]: [number, number]) => a / b })
export const $multiply = operator({ operator: '*', symbol: '$multiply', fn: ([a, b]: [number, number]) => a * b })
export const $modulo = operator({ operator: '%', symbol: '$modulo', fn: ([a, b]: [number, number]) => a % b })
