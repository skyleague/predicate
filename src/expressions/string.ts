import { operator } from '../engine/operator.js'

export const $concat = operator({
    operator: 'concat',
    symbol: '$concat',
    fn: ([a, b, ...xs]: [string, string, ...string[]]) => a.concat(b, ...xs),
})
