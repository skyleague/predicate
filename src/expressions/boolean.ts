import { inspect } from 'node:util'
import { equal } from '@skyleague/axioms'
import { type FactsFomExprs, operator } from '../engine/operator.js'
import {
    type AsExpression,
    type Expression,
    type ExpressionTypeOfLiteral,
    fromLiteral,
    type InputFromExpressions,
    type LiteralOr,
    type ValueExpression,
} from '../engine/types.js'
import { $value, type ValueItem } from './higher-order-fn.js'

export const $startsWith = operator({
    operator: 'startsWith',
    symbol: '$startsWith',
    fn: ([str, searchString]: [str: string, searchString: string]) => str.startsWith(searchString),
})
export const $endsWith = operator({
    operator: 'endsWith',
    symbol: '$endsWith',
    fn: ([str, searchString]: [str: string, searchString: string]) => str.endsWith(searchString),
})
export const $includes = operator({
    operator: 'includes',
    symbol: '$includes',
    fn: ([str, searchString]: [str: string, searchString: string]) => str.includes(searchString),
})

export const $equal = operator({ operator: '==', symbol: '$equal', fn: ([a, b]: [unknown, unknown]) => equal(a, b) })

export const $not = operator({ operator: '~', symbol: '$not', fn: ([x]: [boolean]) => !x })

export const $gt = operator({ operator: '>', symbol: '$gt', fn: ([a, b]: [number, number]) => a > b })
export const $gte = operator({ operator: '>=', symbol: '$gte', fn: ([a, b]: [number, number]) => a >= b })
export const $lt = operator({ operator: '<', symbol: '$lt', fn: ([a, b]: [number, number]) => a < b })
export const $lte = operator({ operator: '<=', symbol: '$lte', fn: ([a, b]: [number, number]) => a <= b })

export const $all = Object.assign(
    // biome-ignore lint/suspicious/noExplicitAny: this is needed for greedy matching
    <Expr extends LiteralOr<any[]>, Out extends Expression<boolean>>(
        xs: Expr,
        predicate: (value: ValueItem<ExpressionTypeOfLiteral<Expr>>) => Out,
    ): ValueExpression<boolean, InputFromExpressions<[AsExpression<Expr>]>, FactsFomExprs<Expr | Out>> => {
        const _xs = fromLiteral(xs)
        const _value = $value<ExpressionTypeOfLiteral<Expr>>()
        const _transform = predicate(_value)
        return {
            fn: ([ys], ctx) => {
                return ys.every((x) => ctx.withScope(x, () => ctx.evaluate(_transform)))
            },
            dependsOn: [_xs],
            expr: (mode) => ({
                all: [_xs.expr(mode), _transform.expr(mode)],
            }),
            [inspect.custom]() {
                return `$all(${_xs[inspect.custom]?.() ?? ''}, (x) => ${_transform[inspect.custom]?.() ?? ''})`
            },
        } as ValueExpression<boolean, InputFromExpressions<[AsExpression<Expr>]>, FactsFomExprs<Expr | Out>>
    },
    // biome-ignore lint/suspicious/noExplicitAny: this is needed for greedy matching
    { operator: 'all', symbol: '$all', parse: (xs: any, predicate: any) => $all(xs, () => predicate) } as const,
)

export const $any = Object.assign(
    // biome-ignore lint/suspicious/noExplicitAny: this is needed for greedy matching
    <Expr extends LiteralOr<any[]>, Out extends Expression<boolean>>(
        xs: Expr,
        predicate: (value: ValueItem<ExpressionTypeOfLiteral<Expr>>) => Out,
    ): ValueExpression<boolean, InputFromExpressions<[AsExpression<Expr>]>, FactsFomExprs<Expr | Out>> => {
        const _xs = fromLiteral(xs)
        const _value = $value<ExpressionTypeOfLiteral<Expr>>()
        const _transform = predicate(_value)
        return {
            fn: ([ys], ctx) => {
                return ys.some((x) => ctx.withScope(x, () => ctx.evaluate(_transform)))
            },
            dependsOn: [_xs],
            expr: (mode) => ({
                any: [_xs.expr(mode), _transform.expr(mode)],
            }),
            [inspect.custom]() {
                return `$any(${_xs[inspect.custom]?.() ?? ''}, (x) => ${_transform[inspect.custom]?.() ?? ''})`
            },
        } as ValueExpression<boolean, InputFromExpressions<[AsExpression<Expr>]>, FactsFomExprs<Expr | Out>>
    },
    // biome-ignore lint/suspicious/noExplicitAny: this is needed for greedy matching
    { operator: 'any', symbol: '$any', parse: (xs: any, predicate: any) => $any(xs, () => predicate) } as const,
)
