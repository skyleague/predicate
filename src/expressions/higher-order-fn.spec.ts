import { $equal } from './boolean.js'
import { $filter, $map, $value, type ValueItem } from './higher-order-fn.js'
import { $fact, $literal, type Fact } from './input.js'
import { $from, type From } from './input.js'
import { $add, $modulo, $multiply } from './number.js'
import { $concat } from './string.js'

import { MathFn } from '../../test/arithmatic.type.js'
import type { Arithmetic } from '../../test/arithmatic.type.js'
import { $policy } from '../engine/policy.js'
import type { Expression, LiteralExpression, ValueExpression } from '../engine/types.js'
import type { JSONExpr, NumberArrExpr, NumberExpr } from '../json/jsonexpr.type.js'

import { array, float, forAll } from '@skyleague/axioms'
import { arbitrary } from '@skyleague/therefore'
import { describe, expect, expectTypeOf, it } from 'vitest'

describe('value', () => {
    it('handles the types', () => {
        const x = $value<number[]>()
        expectTypeOf(x).toEqualTypeOf<ValueItem<number[]>>()
        const _y: Expression = x
        void _y
    })
})

describe('map', () => {
    it('handles a simple schema case', () => {
        const input = {
            a: [1, 2, 3],
            b: [],
        }
        const fact = $fact(MathFn, 'input')
        const x1 = $from(fact, '$.a')
        const x2 = $from(fact, '$.b')
        const x3 = $map(x1, (x) => $add(x, 1))
        const hof = $policy({ x1, x2, x3 })
        const result = hof.evaluate({ input })

        expect(result).toMatchInlineSnapshot(`
          {
            "input": {
              "input": {
                "a": [
                  1,
                  2,
                  3,
                ],
                "b": [],
              },
            },
            "output": {
              "x1": [
                1,
                2,
                3,
              ],
              "x2": [],
              "x3": [
                2,
                3,
                4,
              ],
            },
          }
        `)
        expectTypeOf(result.input.input).toEqualTypeOf<MathFn>()
        expectTypeOf(result.output).toEqualTypeOf<{ x1: number[]; x2: Arithmetic[]; x3: number[] }>()
        expect(hof.expr()).toMatchSnapshot()
    })

    it('handles any simple schema case', () => {
        const fact = $fact(MathFn, 'input')
        const x1 = $from(fact, '$.a')
        const x2 = $from(fact, '$.b[*].b')
        const b = $from(fact, '$.b')
        const x3 = $map(x1, (x) => $add(x, 1))
        const x4 = $map(x2, (x) => $multiply(2, x))
        const x5 = $map(b, (x) => $multiply(2, x('$.b')))
        const hof = $policy({ x1, x2, x3, x4, x5 })
        forAll(arbitrary(MathFn), (input) => {
            const result = hof.evaluate({ input })
            expect(result.input).toEqual({ input })
            expect(result.output).toEqual({
                x1: input.a,
                x2: input.b.map((a) => a.b),
                x3: input.a.map((a) => a + 1),
                x4: input.b.map((a) => a.b * 2),
                x5: input.b.map((a) => a.b * 2),
            })
            expectTypeOf(result.input.input).toEqualTypeOf<MathFn>()
            expectTypeOf(result.output).toEqualTypeOf<{
                x1: number[]
                x2: number[]
                x3: number[]
                x4: number[]
                x5: number[]
            }>()
        })
    })

    it('handles primitive', () => {
        const x1 = $map([1, 2], (x) => $add(x, 1))
        const x2 = $map(['1', '2'], (x) => $concat(x, '1'))
        const x3 = $map([true, false], (x) => x)
        const x4 = $map([1, 2], () => $literal([1, 2]))
        const x5 = $map(['1', '2'], () => $literal(['1', '2']))
        const x6 = $map([true, false], () => $literal([true, false]))
        const hof = $policy({ x1, x2, x3, x4, x5, x6 })
        const result = hof.evaluate()
        expect(result).toMatchInlineSnapshot(`
          {
            "input": undefined,
            "output": {
              "x1": [
                2,
                3,
              ],
              "x2": [
                "11",
                "21",
              ],
              "x3": [
                true,
                false,
              ],
              "x4": [
                [
                  1,
                  2,
                ],
                [
                  1,
                  2,
                ],
              ],
              "x5": [
                [
                  "1",
                  "2",
                ],
                [
                  "1",
                  "2",
                ],
              ],
              "x6": [
                [
                  true,
                  false,
                ],
                [
                  true,
                  false,
                ],
              ],
            },
          }
        `)
        expectTypeOf(result.output).toEqualTypeOf<{
            x1: number[]
            x2: string[]
            x3: boolean[]
            x4: number[][]
            x5: string[][]
            x6: boolean[][]
        }>()
        expect(hof.expr()).toMatchSnapshot()
    })

    it('handles any numbers', () => {
        forAll(array(float()), (xs) => {
            const x1 = $map(xs, (x) => $add(x, 1))
            const hof = $policy({ x1 })
            const result = hof.evaluate()
            expect(result.output.x1).toEqual(xs.map((x) => x + 1))
        })
    })

    it('handles the types', () => {
        const fact = $fact(MathFn, 'input')

        const x1 = $map([1, 2], (x) => x)
        expectTypeOf(x1).toEqualTypeOf<ValueExpression<number[], [LiteralExpression<number[], NumberArrExpr>], NumberExpr>>()

        const a = $from(fact, '$.a')
        const x2 = $map(a, (x) => x)
        expectTypeOf(x2).toEqualTypeOf<ValueExpression<number[], [From<MathFn, number[], [Fact<MathFn, 'input'>]>], NumberExpr>>()

        const ba = $from(fact, '$.b..a')
        const x3 = $map(ba, (x) => x)
        expectTypeOf(x3).toEqualTypeOf<ValueExpression<number[], [From<MathFn, number[], [Fact<MathFn, 'input'>]>], NumberExpr>>()

        const b = $from(fact, '$.b')
        const x4 = $map(b, (x) => x('$.a'))
        expectTypeOf(x4).toEqualTypeOf<
            ValueExpression<number[], [From<MathFn, Arithmetic[], [Fact<MathFn, 'input'>]>], NumberExpr>
        >()
    })
})

describe('filter', () => {
    it('handles a simple schema case', () => {
        const input = {
            a: [1, 2, 3],
            b: [],
        }
        const fact = $fact(MathFn, 'input')
        const x1 = $from(fact, '$.a')
        const x2 = $from(fact, '$.b')
        const x3 = $filter(x1, (x) => $equal(0, $modulo(x, 2)))
        const hof = $policy({ x1, x2, x3 })
        const result = hof.evaluate({ input })

        expect(result).toMatchInlineSnapshot(`
          {
            "input": {
              "input": {
                "a": [
                  1,
                  2,
                  3,
                ],
                "b": [],
              },
            },
            "output": {
              "x1": [
                1,
                2,
                3,
              ],
              "x2": [],
              "x3": [
                2,
              ],
            },
          }
        `)
        expectTypeOf(result.input.input).toEqualTypeOf<MathFn>()
        expectTypeOf(result.output).toEqualTypeOf<{ x1: number[]; x2: Arithmetic[]; x3: number[] }>()
        expect(hof.expr()).toMatchSnapshot()
    })

    it('handles any simple schema case', () => {
        const fact = $fact(MathFn, 'input')
        const x1 = $from(fact, '$.a')
        const x2 = $from(fact, '$.b[*].b')
        const b = $from(fact, '$.b')
        const x3 = $filter(x1, (x) => $equal(0, $modulo(x, 2)))
        const x4 = $filter(x2, (x) => $equal(1, $modulo(x, 2)))
        const x5 = $filter(b, (x) => $equal(1, $modulo(x('$.b'), 2)))
        const hof = $policy({ x1, x2, x3, x4, x5 })
        forAll(arbitrary(MathFn), (input) => {
            const result = hof.evaluate({ input })
            expect(result.input).toEqual({ input })
            expect(result.output).toEqual({
                x1: input.a,
                x2: input.b.map((a) => a.b),
                x3: input.a.filter((a) => a % 2 === 0),
                x4: input.b.filter((a) => a.b % 2 === 1),
                x5: input.b.filter((a) => a.b % 2 === 1),
            })
            expectTypeOf(result.input.input).toEqualTypeOf<MathFn>()
            expectTypeOf(result.output).toEqualTypeOf<{
                x1: number[]
                x2: number[]
                x3: number[]
                x4: number[]
                x5: Arithmetic[]
            }>()
        })
    })

    it('handles any numbers', () => {
        forAll(array(float()), (xs) => {
            const x1 = $filter(xs, (x) => $equal($modulo(x, 2), 1))
            const hof = $policy({ x1 })
            const result = hof.evaluate()
            expect(result.output.x1).toEqual(xs.filter((x) => x % 2 === 1))
        })
    })

    it('handles the types', () => {
        const fact = $fact(MathFn, 'input')

        const x1 = $filter([1, 2], (x) => $equal(x, 1))
        expectTypeOf(x1).toEqualTypeOf<ValueExpression<number[], [LiteralExpression<number[], NumberArrExpr>], NumberArrExpr>>()

        const a = $from(fact, '$.a')
        const x2 = $filter(a, (x) => $equal(x, 1))
        expectTypeOf(x2).toEqualTypeOf<
            ValueExpression<number[], [From<MathFn, number[], [Fact<MathFn, 'input'>]>], NumberArrExpr>
        >()

        const ba = $from(fact, '$.b..a')
        const x3 = $filter(ba, (x) => $equal(x, 1))
        expectTypeOf(x3).toEqualTypeOf<
            ValueExpression<number[], [From<MathFn, number[], [Fact<MathFn, 'input'>]>], NumberArrExpr>
        >()

        const b = $from(fact, '$.b')
        const x4 = $filter(b, (x) => $equal(1, x('$.a')))
        expectTypeOf(x4).toEqualTypeOf<
            ValueExpression<Arithmetic[], [From<MathFn, Arithmetic[], [Fact<MathFn, 'input'>]>], JSONExpr>
        >()
    })
})
