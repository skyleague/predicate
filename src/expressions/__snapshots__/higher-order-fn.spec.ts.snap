// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`filter > handles a simple schema case 2`] = `
{
  "input": {
    "input": {
      "$defs": {
        "Arithmetic": {
          "additionalProperties": true,
          "properties": {
            "a": {
              "type": "number",
            },
            "b": {
              "type": "number",
            },
          },
          "required": [
            "a",
            "b",
          ],
          "title": "Arithmetic",
          "type": "object",
        },
      },
      "$schema": "http://json-schema.org/draft-07/schema#",
      "additionalProperties": true,
      "properties": {
        "a": {
          "items": {
            "type": "number",
          },
          "type": "array",
        },
        "b": {
          "items": {
            "$ref": "#/$defs/Arithmetic",
          },
          "type": "array",
        },
      },
      "required": [
        "a",
        "b",
      ],
      "title": "MathFn",
      "type": "object",
    },
  },
  "meta": {
    "version": "1.0.0",
  },
  "output": {
    "x1": {
      "from": [
        "input",
        "$.a",
      ],
    },
    "x2": {
      "from": [
        "input",
        "$.b",
      ],
    },
    "x3": {
      "filter": [
        {
          "from": [
            "input",
            "$.a",
          ],
        },
        {
          "==": [
            0,
            {
              "%": [
                {
                  "value": "$",
                },
                2,
              ],
            },
          ],
        },
      ],
    },
  },
}
`;

exports[`map > handles a simple schema case 2`] = `
{
  "input": {
    "input": {
      "$defs": {
        "Arithmetic": {
          "additionalProperties": true,
          "properties": {
            "a": {
              "type": "number",
            },
            "b": {
              "type": "number",
            },
          },
          "required": [
            "a",
            "b",
          ],
          "title": "Arithmetic",
          "type": "object",
        },
      },
      "$schema": "http://json-schema.org/draft-07/schema#",
      "additionalProperties": true,
      "properties": {
        "a": {
          "items": {
            "type": "number",
          },
          "type": "array",
        },
        "b": {
          "items": {
            "$ref": "#/$defs/Arithmetic",
          },
          "type": "array",
        },
      },
      "required": [
        "a",
        "b",
      ],
      "title": "MathFn",
      "type": "object",
    },
  },
  "meta": {
    "version": "1.0.0",
  },
  "output": {
    "x1": {
      "from": [
        "input",
        "$.a",
      ],
    },
    "x2": {
      "from": [
        "input",
        "$.b",
      ],
    },
    "x3": {
      "map": [
        {
          "from": [
            "input",
            "$.a",
          ],
        },
        {
          "+": [
            {
              "value": "$",
            },
            1,
          ],
        },
      ],
    },
  },
}
`;

exports[`map > handles primitive 2`] = `
{
  "input": {},
  "meta": {
    "version": "1.0.0",
  },
  "output": {
    "x1": {
      "map": [
        [
          1,
          2,
        ],
        {
          "+": [
            {
              "value": "$",
            },
            1,
          ],
        },
      ],
    },
    "x2": {
      "map": [
        [
          "1",
          "2",
        ],
        {
          "concat": [
            {
              "value": "$",
            },
            "1",
          ],
        },
      ],
    },
    "x3": {
      "map": [
        [
          true,
          false,
        ],
        {
          "value": "$",
        },
      ],
    },
    "x4": {
      "map": [
        [
          1,
          2,
        ],
        [
          1,
          2,
        ],
      ],
    },
    "x5": {
      "map": [
        [
          "1",
          "2",
        ],
        [
          "1",
          "2",
        ],
      ],
    },
    "x6": {
      "map": [
        [
          true,
          false,
        ],
        [
          true,
          false,
        ],
      ],
    },
  },
}
`;
